ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -799,8 +800,12 @@ func (r *RayClusterReconciler) buildWorkerPod(instance rayiov1alpha1.RayCluster,
 // SetupWithManager builds the reconciler.
 func (r *RayClusterReconciler) SetupWithManager(mgr ctrl.Manager, reconcileConcurrency int) error {
 	b := ctrl.NewControllerManagedBy(mgr).
-		For(&rayiov1alpha1.RayCluster{}).Named("raycluster-controller").
-		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{}, predicate.AnnotationChangedPredicate{})).
+		Named("raycluster-controller").
+		For(&rayiov1alpha1.RayCluster{}, builder.WithPredicates(predicate.Or(
```

Comment:
Based on [this code snippet](https://stackoverflow.com/a/71649836/553994). I couldn't find a way to test this since I don't know how to force a change to a Pod's status.

Reply:
There may be a better way, but one manual test would be the script at the bottom of the PR description here https://github.com/ray-project/kuberay/pull/850:
The head pod naturally changes from PENDING -> RUNNING.  Without that PR, the cluster doesn't reconcile upon this status change, so the job never gets submitted.  (With that PR, it would reconcile every 60 seconds.)

With your PR, job would get submitted as soon as the cluster was ready.

```
      # This script is used to rebuild the KubeRay operator image and load it into the Kind cluster. 
# Step 0: Delete the existing Kind cluster
kind delete cluster 

# Step 1: Create a Kind cluster
kind create cluster --image=kindest/node:v1.24.0

# Step 2: Modify KubeRay source code
# For example, add a log "Hello KubeRay" in the function `Reconcile` in `raycluster_controller.go`.

# Step 3: Build a Docker image
#         This command will copy the source code directory into the image, and build it.
# Command: IMG={IMG_REPO}:{IMG_TAG} make docker-build
IMG=kuberay/operator:nightly make docker-build

# Step 4: Load the custom KubeRay image into the Kind cluster.
# Command: kind load docker-image {IMG_REPO}:{IMG_TAG}
kind load docker-image kuberay/operator:nightly

# Step 5: Keep consistency
# If you update RBAC or CRD, you need to synchronize them.
# See the section "Consistency check" for more information.

# Step 6: Install KubeRay operator with the custom image via local Helm chart
# (Path: helm-chart/kuberay-operator)
# Command: helm install kuberay-operator --set image.repository={IMG_REPO} --set image.tag={IMG_TAG} .
helm install kuberay-operator --set image.repository=kuberay/operator --set image.tag=nightly ~/kuberay/helm-chart/kuberay-operator

# Step 7: Submit a RayJob
kubectl apply -f config/samples/ray_v1alpha1_rayjob.yaml   

# Step 8: Check the status of the pods
watch kubectl get pods

# When the pods are ready, check `kubectl logs` to see that the job has status SUCCEEDED
```

Reply:
I agree the best way to test this is by verifying that it fixes the RayJob issue!

Reply:
cc @sihanwang41 we probably want to make the same adjustment to the RayService controller --
I remember you asked if there was a difference between the "event filter" and "predicate" arrangement -- we figured there wasn't a difference and went with "event filter"

Reply:
Also, ideally we should add a unit test for this (should be possible to follow the pattern of one of the existing tests)

Reply:
@DmitriGekhtman should the unit test be added to [raycluster_controller_test.go](https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/raycluster_controller_test.go)? There's a test there check that the [controller recreates a deleted worker Pod](https://github.com/ray-project/kuberay/blob/2600854c61673f2b7da9fe2b54c8220468c1a013/ray-operator/controllers/ray/raycluster_controller_test.go#L193-L208). This test currently passes on master without this change. Is this expected? I guess globalPredicates don't filter out deletion events of owned resources?

If so, I'm thinking we can write a test that creates an event on a child resource that's currently filtered out by the globalPredicates like a Pod status change. Then we check if the RayCluster's `.status.lastUpdateTime` is updated? If this sounds good, any ideas on the test code to write to force a Pod status change that will currently fail when run against master branch?

I also updated `rayservice_controller.go`.

Reply:
It's expected that the test passes -- the current problem is that pod status changes are ignored.
I think it should be possible to use the machinery of that test to verify that RayCluster status.state transitions to Ready within half a second of all managed pods transitioning to Running.

Reply:
Thanks, added a test. Lmk how it looks

Reply:
looks good